adapters/beintoo/beintoo.go

```diff
@@ -0,0 +1,246 @@
+}
+
+func buildEndpoint(endpoint string, testing bool, timeout int64) string {
+	if timeout == 0 {
+		timeout = 1000
+	}
+	if testing {
+		// for passing validation tests
+		return endpoint + "?t=1000&ts=2060541160"
+	}
+	return endpoint + "?t=" + strconv.FormatInt(timeout, 10) + "&ts=" + strconv.FormatInt(time.Now().Unix(), 10) + "&src=pbserver"
+}
+
+func (a *BeintooAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("No Imps in Bid Request"),
+		}}
+	}
+
+	if errs := preprocess(request); errs != nil && len(errs) > 0 {
+		return nil, append(errs, &errortypes.BadInput{
+			Message: fmt.Sprintf("Error in preprocess of Imp, err: %s", errs),
+		})
+	}
+
+	data, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Error in packaging request to JSON"),
+		}}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	if request.Device != nil {
+		addHeaderIfNonEmpty(headers, "User-Agent", request.Device.UA)
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+	if request.Site != nil {
+		addHeaderIfNonEmpty(headers, "Referer", request.Site.Page)
+	}
+
+	url := buildEndpoint(a.endpoint, a.testing, request.TMax)
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     url,
+		Body:    data,
+		Headers: headers,
+	}}, errs
+}
+
+func unpackImpExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpBeintoo, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var BeintooExt openrtb_ext.ExtImpBeintoo
+	if err := json.Unmarshal(bidderExt.Bidder, &BeintooExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid ImpExt", imp.ID),
+		}
+	}
+
+	tagIDValidation, err := strconv.ParseInt(BeintooExt.TagID, 10, 64)
+	if err != nil || tagIDValidation == 0 {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid tagid must be a String of numbers", imp.ID),
+		}
+	}
+
+	if BeintooExt.TagID == "" {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Ignoring imp id=%s, no tagid present", imp.ID),
+		}
+	}
+
+	return &BeintooExt, nil
+}
+
+func buildImpBanner(imp *openrtb.Imp) error {
+	imp.Ext = nil
+
+	if imp.Banner == nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Request needs to include a Banner object"),
+		}
+	}
+
+	bannerCopy := *imp.Banner
+	banner := &bannerCopy
```

Comment:
I don't think we need the `bannerCopy` because the `buildImpBanner` already was passed an imp that is a copy, a copy created by the `for` statement in line 181 which is a copy valid in its scope. It doesn't hurt to keep it there, but it'd be a small optimization.

Reply:
I'm sorry @ddantuonobeintoo you were right. This copy is needed. I'm sorry, can we rollback to use it again?